### ISSUE TITLE
Creating repeater item via API is inconsistent depending on if adding to newly created page vs existing page?

### ISSUE THREAD

**Issue by jlahijani** [processwire/processwire-issues#1832]
I think this is a bug unless I'm just not considering certain practices.  I don't know if this is an issue that started to occur in the last 6 months or just me coding something that I haven't done before, but I went deep to find this potential issue.

I'm using bleeding edge ProcessWire and PHP 8.1.

Assume you have a basic-page with a repeater field called "books".

Let's say we want to create a page and add an item to it in a script located at `/scripts/create-page-and-add-repeater-item.php`:

```
<?php namespace ProcessWire;
include(dirname(__FILE__).'/../index.php');

// create page
$p = new Page();
$p->template = 'basic-page';
$p->parent = $pages->get('/');
$p->name = 'basic-page-1';
$p->title = 'Basic Page 1';
$p->save();

// create repeater item
$book = $p->books->getNewItem();
$book->title = 'book item 1';
$book->save();
$p->save();

echo $p->books->count; // correctly outputs 1
echo $pages->find('template=basic-page,books.count=1')->count; // correctly outputs 1
```

Now while the above outputs the correct information, my understanding is that I am supposed to add the freshly created repeater item to the books repeater explicitly.  Here's the same script but with that addition, however notice the incorrect output below:

```
<?php namespace ProcessWire;
include(dirname(__FILE__).'/../index.php');

// create page
$p = new Page();
$p->template = 'basic-page';
$p->parent = $pages->get('/');
$p->name = 'basic-page-1';
$p->title = 'Basic Page 1';
$p->save();

// create repeater item
$book = $p->books->getNewItem();
$book->title = 'book item 1';
$book->save();
$p->books->add($book); // adding book explicitly
$p->save();

echo $p->books->count; // incorrectly outputs 0
echo $pages->find('template=basic-page,books.count=1')->count; // incorrectly outputs 0
```

After a lot of debugging, what I realized is there's perhaps some bug where doing `$p->books->add($book);` in this specific scenario (meaning adding a repeater item right after a page was created in the same script) adds the repeater item twice (please use TracyDebugger to inspect it along the way).

To counteract (and what seems like a hack to overcome this bug), I am changing that line from this:
```
$p->books->add($book); 
```

To this:
```
if(!$p->get('_added')) $p->books->add($book); // this doesn't execute in this scenario since page was just added
```

It then correctly outputs the counts at the end.

---

*NOW DELETE THAT TEST PAGE COMPLETELY AND CONTINUE SO YOU DON'T MIX UP YOUR TESTING*

---

So that fixes the issue when creating a page then adding a repeater immediately afterwards.

However there's also an issue when creating a page, then adding a repeater later.

Let's create a page in `/scripts/create-page.php`:
```
<?php namespace ProcessWire;
include(dirname(__FILE__).'/../index.php');

// create page
$p = new Page();
$p->template = 'basic-page';
$p->parent = $pages->get('/');
$p->name = 'basic-page-1';
$p->title = 'Basic Page 1';
$p->save();
```

Now let's add a repeater in `/scripts/add-repeater-item.php`:

```
<?php namespace ProcessWire;
include(dirname(__FILE__).'/../index.php');

// get the page that was previously created right above in /scripts/create-page.php
$p = $pages->get('name=test-book-1');

// add repeater item to page
$book = $p->books->getNewItem();
$book->title = 'book item 1';
$book->save();
$p->books->add($book); // adding book explicitly
$p->save();

echo $p->books->count; // correctly outputs 1
echo $pages->find('template=basic-page,books.count=1')->count; // correctly outputs 1
```

Notice how the output was correct despite explicitly adding `$p->books->add($book);`.

If I do not have `$p->books->add($book);` here's what happens:

```
<?php namespace ProcessWire;
include(dirname(__FILE__).'/../index.php');

// get the page that was previously created right above in /scripts/create-page.php
$p = $pages->get('name=test-book-1');

// add repeater item to page
$book = $p->books->getNewItem();
$book->title = 'book item 1';
$book->save();
$p->save();

echo $p->books->count; // correctly outputs 1
echo $pages->find('template=basic-page,books.count=1')->count; // incorrectly outputs 0
```

So again, to make the adding of the repeater feel *consistent* regardless of whether the page was just added or not, I could modify that one line to be like this:
```
if(!$p->get('_added')) $p->books->add($book); // this DOES execute in this scenario since page existed before
```

---

In short, shouldn't doing `$p->books->add($book);` whether it's a new page or existing page yield the same results?

**Comment by jlahijani** [comment]
Would it make sense to add this line to prevent duplicates in `FieldtypeRepeater.module`?

```
// iterate through the array and count how many published we have
foreach($value as $p) {
	$numTotal++;
	if(!$p->id || $p->isHidden() || $p->isUnpublished()) continue; 
	if(in_array($p->id, $ids)) continue; // !!! prevent duplicates !!!
	$ids[] = $p->id;
	$numPublished++;
}
```

**Comment by BernhardBaumrock** [comment]
I'm seeing the same issue! But your fix does not work unfortunately @jlahijani and I didn't find it myself :(

@ryancramerdesign would be great if you can have a look :)

**Comment by jlahijani** [comment]
Thanks for confirming @BernhardBaumrock .  I haven't fully tested my fix to prevent duplicates yet, but I will today.

**Comment by jlahijani** [comment]
@BernhardBaumrock From what I can tell, my fix of adding `if(in_array($p->id, $ids)) continue;` works.  Ignore all the stuff I posted in my first post, meaning this should not be done:
`if(!$p->get('_added')) $p->books->add($book);`

But instead this should as it's the correct way according to the API:
`$p->books->add($book);`

Does doing it the way I mentioned work for you?  Do you have a set of steps I can test if you're still having issues?

**Comment by BernhardBaumrock** [comment]
Hey @jlahijani I have a more complicated setup with RockPageBuilder, so unfortunately it's not as easy as in your case. I'll have to dig deeper into that. I was hoping that your fix also fixes my issue in RockPageBuilder :)

**Comment by ryancramerdesign** [comment]
@jlahijani @BernhardBaumrock Thanks, I was able to duplicate and have pushed a fix for this, though want to confirm it also fixes it for you? The `$p->books->add($book)` shouldn't be necessary, as `getNewItem()` already does an add(). From what I could tell, the issue was more that the change doesn't get tracked on the page that contains the repeater because the item being added doesn't yet have an ID. So when the $p page is saved, it doesn't see the change to the repeater and thus doesn't save the field, which is where the count gets updated. I added a `$p->trackChange('repeater_field')` in the `getNewItem()` method and that seems to fix it here at least. Can you confirm? Thanks.

**Comment by jlahijani** [comment]
@ryancramerdesign It didn't seem to resolve the issue and there are some quirks.  Since this is a bit complicated to communicate, I made a video which I believe will help:
https://www.youtube.com/watch?v=76_gffLviws

Here's the testing script as seen in the video:
```
<?php namespace ProcessWire;
/*
http://processwire-test.test/admin/
//
http://processwire-test.test/scripts/x1.php?action=create-page
http://processwire-test.test/scripts/x1.php?action=create-repeater-item
//
http://processwire-test.test/scripts/x1.php?action=create-page-and-create-repeater-item
//
http://processwire-test.test/scripts/x1.php?action=count
http://processwire-test.test/scripts/x1.php?action=edit
http://processwire-test.test/scripts/x1.php?action=delete
*/
include(dirname(__FILE__).'/../index.php');

if($input->action=="create-page") {
  $p = new Page();
  $p->template = 'rptest';
  $p->parent = $pages->get('/');
  $p->name = 'test-book-1';
  $p->title = 'test-book-1';
  $p->save();
} elseif($input->action=="create-repeater-item") {
  $p = $pages->get('name=test-book-1');
  $book = $p->books->getNewItem();
  $book->title = 'test 1';
  $book->save();
  $p->books->add($book);
  $p->save();
} elseif($input->action=="create-page-and-create-repeater-item") {
  $p = new Page();
  $p->template = 'rptest';
  $p->parent = $pages->get('/');
  $p->name = 'test-book-1';
  $p->title = 'test-book-1';
  $p->save();
  $book = $p->books->getNewItem();
  $book->title = 'test 1';
  $book->save();
  $p->books->add($book);
  $p->save();
} elseif($input->action=="count") {
  $p = $pages->get('name=test-book-1');
  echo $p->books->count;
  echo "<br>";
  echo $pages->find('template=rptest,books.count=1')->count;
} elseif($input->action=="edit") {
  $p = $pages->get('name=test-book-1');
  $session->redirect($p->editUrl);
} elseif($input->action=="delete") {
  foreach($pages->find('template=rptest,include=all') as $p) {
    $p->delete(true);
  }
  $session->redirect("/admin/");
}

```

**Comment by ryancramerdesign** [comment]
@jlahijani It says the video is private. But before we look at other things, let's stay on that original reported issue, since that's the one I was able to test and and clearly see here. I was able to fix what I saw here at least, so would like to identify why it didn't fix there. You'll want to delete any existing "test 1" items in the repeater before testing. 

I think the closest would be your  'create-repeater-item' action. But since you created the item with getNewItem(), the `$p->books->add()` line is not necessary. So if you remove it, that'd leave you with this:
~~~~~
$p = $pages->get('name=test-book-1');
$book = $p->books->getNewItem();
$book->title = 'test 1';
$book->save();
$p->save();
~~~~~
Prior to the committed fix, it would save a count of 0 for the repeater field in the DB. So the repeater item would exist and get loaded if `$p->books` was accessed, but you couldn't query it with `count=...` in a $pages->find/get selector. 

Following the fix, items added to the books field (from this point forward) should be updated with the count, so that a find/get on the count property should now work, at least once a new item is added and saved. Though stick with a `count>0` so that you don't have to delete all the repeater items on every test. 
~~~~~
$p = $pages->get("name=test-book-1, books.count>0");
if($p->id) {
  echo "Found with count=" . $p->books->count();
} else {
  echo "Not found";
}
~~~~~

**Comment by jlahijani** [comment]
Sorry about that.  I fixed the video to be viewable.

**Comment by jlahijani** [comment]
Actually I can confirm the original fix does work correctly.  Thanks!
